### ISSUE TITLE
web-1  | nginx: [warn] the "listen ... http2" directive is deprecated, use the "http2" directive instead

### ISSUE THREAD

**Issue by replete** [shinsenter/php#74]
# Main Issue:
```
web-1  | nginx: [warn] the "listen ... http2" directive is deprecated, use the "http2" directive instead
```

This deprecation warning can be fixed by moving http2 into a new directive like so:

```conf
server {
    listen 80 reuseport default_server;
    listen [::]:80 reuseport default_server;
    listen 443 ssl reuseport;
    listen [::]:443 ssl reuseport;

    http2 on;
	
	# ...
```

# Secondary query:
Also can you please tell us how best to use these images with custom nginx configuration? This is my workaround for custom nginx configuration:

Currently, I copy the nginx config from the running container:
```docker cp $(docker compose -f ../docker-compose.yml ps -q web):/etc/nginx ../nginx-dev-config```

And then mount the files individually with volumes in my `docker-compose.yml`:
```yaml
    volumes:
      # App
      - ./statamic:/var/www/html
      # Dev SSL keys (npm run devcerts in /statamic)
      - ./dev.crt:/etc/ssl/site/server.crt:ro
      - ./dev.key:/etc/ssl/site/server.key:ro
      # Nginx dev config
      - ./nginx-dev-config/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./nginx-dev-config/sites-enabled/00-default.conf:/etc/nginx/sites-enabled/00-default.conf:ro
      - ./nginx-dev-config/fastcgi.conf:/etc/nginx/fast-cgi.conf:ro
      - ./nginx-dev-config/custom.d/common.conf:/etc/nginx/custom.d/common.conf:ro
```

But this generates these errors in docker:
```
web-1  | sed: can't move '/etc/nginx/sites-enabled/00-default.confgdAIiC' to '/etc/nginx/sites-enabled/00-default.conf': Resource busy
web-1  | sed: can't move '/etc/nginx/nginx.confBAdMAD' to '/etc/nginx/nginx.conf': Resource busy
web-1  | sed: can't move '/etc/nginx/nginx.confidBlHD' to '/etc/nginx/nginx.conf': Resource busy
web-1  | nginx version: nginx/1.26.1
web-1  | sed: can't move '/etc/nginx/custom.d/common.confIfHmAE' to '/etc/nginx/custom.d/common.conf': Resource busy
web-1  | nginx: [warn] "ssl_stapling" ignored, issuer certificate not found for certificate "/etc/ssl/site/server.crt"
web-1  | nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
web-1  | nginx: configuration file /etc/nginx/nginx.conf test is successful
web-1  | [12-Jul-2024 10:46:31] NOTICE: fpm is running, pid 381
web-1  | [12-Jul-2024 10:46:31] NOTICE: ready to handle connections
web-1  | nginx: [warn] "ssl_stapling" ignored, issuer certificate not found for certificate "/etc/ssl/site/server.crt"
```

This image is advertised as production ready, so could you help me understand how you would make custom nginx configurations, is it similar to this? I would prefer to eliminate all warnings and errors from a production environment.

Thank you once again for great work.

**Comment by shinsenter** [comment]
Hi @replete 

Thank you for reaching out.

Regarding the issue "http2 directive is deprecated", it is not urgent at the moment, and nginx can still function normally. Therefore, you can use it without too much concern.

You can add your nginx configurations by mounting or copying your `*.conf` files into the `/etc/nginx/custom.d/` directory of the container. The config files here will be loaded into the default nginx server. Since there are already some common configs in this directory, please avoid mounting or replacing an entire different directory into `/etc/nginx/custom.d/`.

**Comment by shinsenter** [comment]
By the way, these Docker images are created to be “ready for production,” not “immutable or non-customizable,” and I primarily focus on PHP. Web server software is added as a gateway to communicate with the underlying PHP handlers, enhancing request processing efficiency and reducing the load on PHP. Users can fully customize, and I am not responsible for any of their customizations.

**Comment by replete** [comment]
Thank you, I understand it's not urgent right now, but for a production environment I want to prevent potential issues in the future

The errors are because I have mounted the copied conf files as read only, so I can control the nginx configuration from source control outside of the container. Sed is complaining that some copy part of your docker image can't replace the files I've mounted, which is what I want.

I understand you are not responsible for customizations, and thank you for your advice so far. I just wanted to understand if my approach made sense.

**Comment by shinsenter** [comment]
Yes you can enable http3 your self by extending these docker images using a Dockerfile.
